halo2-ecc/src/bn254/tests/bls_signature.rs

```diff
@@ -0,0 +1,235 @@
+use std::{
+    fs::{self, File},
+    io::{BufRead, BufReader},
+};
+
+use super::*;
+use crate::{fields::FpStrategy, halo2_proofs::halo2curves::bn256::G2Affine};
+use halo2_base::{
+    gates::{
+        builder::{
+            CircuitBuilderStage, GateThreadBuilder, MultiPhaseThreadBreakPoints,
+            RangeCircuitBuilder,
+        },
+        RangeChip,
+    },
+    halo2_proofs::{
+        halo2curves::{
+            bn256::{multi_miller_loop, G2Prepared, Gt},
+            pairing::MillerLoopResult,
+        },
+        poly::kzg::multiopen::{ProverGWC, VerifierGWC},
+    },
+    utils::fs::gen_srs,
+    Context,
+};
+use rand_core::OsRng;
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
+struct BlsSignatureCircuitParams {
+    strategy: FpStrategy,
+    degree: u32,
+    num_advice: usize,
+    num_lookup_advice: usize,
+    num_fixed: usize,
+    lookup_bits: usize,
+    limb_bits: usize,
+    num_limbs: usize,
+    num_aggregation: u32,
+}
+
+/// Verify e(g1, signature_agg) = e(pubkey_agg, H(m))
+fn bls_signature_test<F: PrimeField>(
+    ctx: &mut Context<F>,
+    params: BlsSignatureCircuitParams,
+    g1: G1Affine,
+    signatures: &[G2Affine],
+    pubkeys: &[G1Affine],
+    msghash: G2Affine,
+) {
+    // Calculate halo2 pairing by multipairing
+    std::env::set_var("LOOKUP_BITS", params.lookup_bits.to_string());
+    let range = RangeChip::<F>::default(params.lookup_bits);
+    let fp_chip = FpChip::<F>::new(&range, params.limb_bits, params.num_limbs);
+    let pairing_chip = PairingChip::new(&fp_chip);
+    let bls_signature_chip = BlsSignatureChip::new(&fp_chip, &pairing_chip);
+    let result = bls_signature_chip.bls_signature_verify(ctx, g1, signatures, pubkeys, msghash);
+
+    // Calculate non-halo2 pairing by multipairing
+    let signature_g2_prepared = G2Prepared::from(signatures.iter().sum::<G2Affine>());
+    let pubkey_aggregated = pubkeys.iter().sum::<G1Affine>();
```

Comment:
Iterator for affine points is used here

Reply:
since this is only in test, it'd be faster if you could just use a for loop here and remove the halo2_proofs dependence. mainly because there will be some upcoming changes/updates to halo2curves so we want to reduce maintenance